docs/DeveloperGuide.md

```diff
@@ -0,0 +1,242 @@
+            clientImpl="my.language.server.MyLanguageClient"
+            scope="project">
+      <description><![CDATA[
+        Some description written in HTML to display it in LSP consoles and Language Servers settings.
+        ]]>
+      </description>
+    </server>
+  </extensions>
+  
+</idea-plugin>
+```
+
+Once the declaration is done, your server should appear in the LSP console:
+
+![My LanguageServer in LSP Console](./images/MyLanguageServerInLSPConsole.png)
+
+## Declare language mapping with extension point
+
+Once server is defined, you need to associate an IntelliJ language with the `server` defined by the id attribute 
+with `com.redhat.devtools.lsp4ij.languageMapping` extension point.
+
+Here a sample which associates the `XML` language with the `myLanguageServerId` server:
+
+```xml
+</idea-plugin>
+
+  <extensions defaultExtensionNs="com.redhat.devtools.lsp4ij">
+
+    <languageMapping language="XML"
+                     serverId="myLanguageServerId" />
+
+  </extensions>
+```
+
+If the language check is not enough, you can implement a [DocumentMatcher](https://github.com/redhat-developer/lsp4ij/blob/main/src/main/java/com/redhat/devtools/lsp4ij/DocumentMatcher.java) to implement some check in Java code.
+For instance our language server could map `Java` as language and you could implement a DocumentMatcher 
+to check if the module of the file have some Java class in the classpath.
+
+The DocumentMatcher is executed in a non blocking read action.
+
+A document matcher looks like this:
+
+```java
+package my.language.server;
+
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.redhat.devtools.lsp4ij.AbstractDocumentMatcher;
+import org.jetbrains.annotations.NotNull;
+
+public class MyDocumentMatcher extends AbstractDocumentMatcher {
+
+    @Override
+    public boolean match(@NotNull VirtualFile virtualFile, @NotNull Project project) {
+        return true;
+    }
+}
+```
+
+and it must be registered in language mapping with `documentMatcher` attribute:
+
+```xml
+</idea-plugin>
+
+  <extensions defaultExtensionNs="com.redhat.devtools.lsp4ij">
+
+    <languageMapping language="XML"
+                     serverId="myLanguageServerId"
+                     documentMatcher="my.language.server.MyDocumentMatcher" />
+
+  </extensions>
+```
+
+## Declare some IJ features
+
+When it is possible, LSP4IJ declare the IJ feature with `any` language 
+For instance `textDocument/publishDiagnostics` is managed with `externalAnnotator` in LSP4IJ with empty language to support any language:
+
+```xml
+<!-- LSP textDocument/publishDiagnostics notification support -->
+<externalAnnotator
+        id="LSPDiagnosticAnnotator"
+        language=""
+        implementationClass="com.redhat.devtools.lsp4ij.operations.diagnostics.LSPDiagnosticAnnotator"/>
+```
+
+Some IJ feature cannot use any language, so you need to declare it in your plugin:

```

Comment:
Some LSP features (Hover, Inlay Hints, CodeLens) have to be mapped specifically to your language, in your plugin.xml: